File: py/aoc2023/day3.py
"""
Day 3: Gear Ratios
"""

import re
from collections import defaultdict
from math import prod

SAMPLE_INPUT = """
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
"""

NUMBER_RE = re.compile(r"\d+")
SYMBOL_RE = re.compile(r"[^\s\d.]")


def _parse(data):
    lines = data.splitlines()
    for y, line in enumerate(lines):
        for match in NUMBER_RE.finditer(line):
            number = int(match.group(0))
            x0, x1 = match.span()
            for yy in range(max(y - 1, 0), min(y + 2, len(lines))):
                for match in SYMBOL_RE.finditer(lines[yy], x0 - 1, x1 + 1):
                    yield match.start(), yy, number


def part1(data):
    """
    >>> part1(SAMPLE_INPUT)
    4361
    """
    return sum(number for _, _, number in _parse(data))


def part2(data):
    """
    >>> part2(SAMPLE_INPUT)
    467835
    """
    gears = defaultdict(list)
    for x, y, number in _parse(data):
        gears[x, y].append(number)
    return sum(prod(numbers) for numbers in gears.values() if len(numbers) == 2)


parts = (part1, part2)

File: hs/src/Day3.hs
{-|
Module:         Day3
Description:    <https://adventofcode.com/2023/day/3 Day 3: Gear Ratios>
-}
module Day3 (part1, part2) where

import Control.Arrow (second)
import Data.Char (isDigit, isSpace)
import qualified Data.Map as Map (elems, fromListWith)
import Data.Text (Text)
import qualified Data.Text as T (break, index, length, lines, span)
import qualified Data.Text.Read as T (decimal)
import qualified Data.Vector as V ((!), fromList, length)

parts :: Text -> [((Int, Int), Int)]
parts input =
  [ ((x, y'), number)
  | y <- [0..V.length lines - 1]
  , ((x0, x1), number) <- numbers 0 $ lines V.! y
  , y' <- [max 0 $ y - 1..min (V.length lines - 1) $ y + 1]
  , let line = lines V.! y'
  , x <- [max 0 $ x0 - 1..min (T.length line - 1) x1]
  , let c = T.index line x
  , not $ c == '.' || isSpace c || isDigit c
  ] where
    lines = V.fromList $ T.lines input
    numbers acc t
      | (leading, t') <- T.break isDigit t
      , Right (number, trailing) <- T.decimal t'
      = let acc' = acc + T.length t - T.length trailing
         in ((acc + T.length leading, acc'), number) : numbers acc' trailing
      | otherwise = []

part1, part2 :: Text -> Int
part1 = sum . map snd . parts
part2 = sum . map product . filter ((== 2) . length) .
    Map.elems . Map.fromListWith (++) . map (second (:[])) . parts
